### PR TITLE
Docs: AutoModeWatcher + SSR

### LINKED ISSUE #2259: Light and dark mode not being derived from os when ssr is disabled
### Current Behavior

I am trying to make my website derive it's theme from the os, I followed the instructions on from the [docs](https://www.skeleton.dev/utilities/lightswitches) but it's just not working and I have no idea why. It's odd because I followed the same steps on another site I'm working on and there it worked without issue.

### Expected Behavior

The theme should match that of the operating system

### Steps To Reproduce

_No response_

### Link to Reproduction / Stackblitz

_No response_

### More Information

_No response_

### PR DESCRIPTION
**Pull request by jwatts777**
## Linked Issue

Closes #2259

## Description

Small changes denoting that a few more steps need to be taken to call `window` +  `document` dependent parts of Lightswitch (autoModeWatcher). 

## Changesets

Instructions: Changesets automate our changelog. If you modify files in `/packages`, run `pnpm changeset` in the root of the monorepo, follow the prompts, then commit the markdown file. Changes that add features should be `minor` while chores and bugfixes should be `patch`. Please prefix the changeset message with `feat:`, `bugfix:` or `chore:`.

## Checklist

Please read and apply all [contribution requirements](https://www.skeleton.dev/docs/contributing).

- [x] This PR targets the `dev` branch (NEVER `master`)
- [x] Documentation reflects all relevant changes
- [x] Branch is prefixed with: `docs/`, `feat/`, `chore/`, `bugfix/`
- [x] Ensure Svelte and Typescript linting is current - run `pnpm ci:check`
- [x] Ensure Prettier linting is current - run `pnpm format`
- [x] All test cases are passing - run `pnpm test`
- [x] Includes a changeset (if relevant; see above)

### CODE REVIEW

**Comment by jwatts777** [comment]
I understand we were using `<svelte:head>` as a workaround for SSR svelte parsing issues, however `<Lightswitch>` has multiple psuedo-`if (browser)` calls of it's own.
 Two actually.
https://github.com/skeletonlabs/skeleton/blob/23683dbece32584954a22d78816857b21395911e/packages/skeleton/src/lib/utilities/LightSwitch/lightswitch.ts#L21
https://github.com/skeletonlabs/skeleton/blob/23683dbece32584954a22d78816857b21395911e/packages/skeleton/src/lib/utilities/LightSwitch/lightswitch.ts#L48
Unsure of how large of a userbase this may serve, however https://github.com/skeletonlabs/skeleton/issues/2259 referenced above received a few quick reports from _newish_ members.
### **_This is not ready to merge, WIP._**

**Comment by endigo9740** [comment]
> I understand we were using <svelte:head> as a workaround for SSR svelte parsing issues, however <Lightswitch> has multiple psuedo-if (browser) calls of it's own.

I do PR reviews on Sundays now, but just to make sure we're on the same page - [browser is a SvelteKit provided module](https://kit.svelte.dev/docs/modules#$app-environment-browser). We cannot utilize that in Skeleton because it would break support for non-Sveltekit projects. While Skeleton only provides technical support for SvelteKit, we need to ensure we're not unintentionally breaking compatibility for Vite or Astro users.

> This is not ready to merge, WIP.

I've gone ahead and marked as a Draft. Bump to "ready for review" when you reach that state please.

**Comment by jwatts777** [comment]
> I do PR reviews on Sundays now, but just to make sure we're on the same page - [browser is a SvelteKit provided module](https://kit.svelte.dev/docs/modules#$app-environment-browser). We cannot utilize that in Skeleton because it would break support for non-Sveltekit projects. While Skeleton only provides technical support for SvelteKit, we need to ensure we're not unintentionally breaking compatibility for Vite or Astro users.

Probably a miscommunication, I have noticed a few inconsistencies in my previous note, but am leaving it up for posterity (and to show that sleep is a good idea).  I was trying to use `if(browser)` to denote the fact that `Lightswitch` also has client-dependent code that _requires_ Window and Document (and included examples above), but `Lightswitch` functions fine with SSR enabled, while `autoModeWatcher` does not.

The issue linked [here](https://github.com/skeletonlabs/skeleton/issues/2259) caught my attention.

This pr is a (tentative) docs change referencing the fact that `autoModeWatcher` has a _requirement_ to be called client side **only** after hydration. Meaning an onMount(); is required to be called so that the `autoModeWatcher` function fires properly, just like it would if injected via `svelte:head` as a script module.

It's all just work arounds to get around how SvelteKit does hydration depending on SSR or not.

In short:

- `Lightswitch` component _**works**_  with `const ssr  = false`
- `autoModeWatcher` does _**not work**_ with `const ssr = false`
- I believe these inconsistencies are related to SK (specifically how svelte:head is treated when `const ssr = false`) and _not_ Skeleton, but we're brushing close to the edge of my expertise.

**Comment by jwatts777** [comment]
I can't really speak for this previous [issue](https://github.com/skeletonlabs/skeleton/issues/905) given that `Lightswitch` functions normally on `const ssr = false` projects.

**Comment by endigo9740** [comment]
@jwatts777 I'm going to leave this PR to review when I have a bit more bandwidth in a couple weeks. Likely around December 18th. Feel free to continue reviewing information and making changes. And again, reminder to set "ready for review" when you reach that point.

**Comment by CallMeLaNN** [comment]
Hello, I'm new, excited and going through this repo, but just to hop in into this because I know a little bit.

I think there's FOUC when put in `onMount()` but it should work as basic example. e.g. try set the user pref to light (since dark is the default) and reload.

The main requirement for the `autoModeWatcher();` that we can tell in the doc is to call or copy it into a `script` tag in `head`. Meaning, it is better to be placed in somewhere similar for Vite or Astro (index.html or layout.astro) without being specific.

File: sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte [hunk 1]
```diff
@@ -60,21 +61,47 @@
...
+									<CodeBlock
+										language="ts"
+										code={`
+import { autoModeWatcher } from '@skeletonlabs/skeleton'; 
+import { onMount } from 'svelte';
+
+onMount(() => {					
+	autoModeWatcher.toString();						
+	autoModeWatcher();
+)}`}
```
[the remark this answers is not in the document]
Comment (jwatts777) [reply sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte:95]: There is no typo. You snipped off `code={`onMount`}`.

File: sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte [hunk 2]
```diff
@@ -60,21 +61,47 @@
...
+									<CodeBlock language="ts" code={`export const ssr = false;`} />
+									<p>Add the following to your root <code class="code">/src/routes/+layout.svelte</code> file</p>
+									<CodeBlock
+										language="ts"
+										code={`
+import { autoModeWatcher } from '@skeletonlabs/skeleton'; 
+import { onMount } from 'svelte';
+
+onMount(() => {					
+	autoModeWatcher.toString();
```
[the remark this answers is not in the document]
Comment (jwatts777) [reply sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte]: I can't tell what you are implying. onMount is required to be called so that `window, document` both exist at call time, otherwise `autoModeWatcher` does not work, and there is no change of styles from light-dark.
If you are referring to the extra `autoModeWatcher.toString()`, that will be removed.

**Comment by jwatts777** [comment]
@CallMeLaNN Please see if you can get FOUC on [this ](https://auto-mode-watcher-test.vercel.app/)
I have noticed none.
If you would like to provide proof for Astro/more, happy to add (or let you add) more to the docs denoting the different paths each framework would benefit from. You would have to provide reproducible repo's, because I frankly don't have an easy way to test edge cases.

File: sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte (hunk 1 above)
[the remark this answers is not in the document]
Comment (CallMeLaNN) [reply sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte:94]: The onMount `)` with fx `}`. `)` should be the last. Not the `code={}`. I look many times and thinking I'm I wrong? 😅 sorry for the trouble.

File: sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte (hunk 2 above)
[the remark this answers is not in the document]
Comment (CallMeLaNN) [reply sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte]: I'm referring exactly on this line only. `toString()` return a string and assigned to nothing. I should be more clear. So this is noop. 

It was for SvelteKit because there's no import in the layout and simply "print" the code of the function definition to be called. I can tell because I just noticed this nice hack in the original layout then found this PR.

[the remark this answers is not in the document]
Comment (jwatts777) [reply sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte]: Understood

File: sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwatts777) [reply sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte:94]: It's ok. The object has to contain a string literal, so the ending } closes off the object `code`. Hope this makes sense

[the remark this answers is not in the document]
Comment (jwatts777) [reply sites/skeleton.dev/src/routes/(inner)/utilities/lightswitches/+page.svelte:94]: Nope, I see what you mean.

**Comment by CallMeLaNN** [comment]
@jwatts777 you are right that it works on your link. I was just guessing from the code.

Being curious how SvelteKit made it, I notice they:

1. Let the head empty (see the original html downloaded) and use `Promise.all` at the end of the body.
2. Inject `<link rel="modulepreload" as="script" crossorigin="" href="...">` to load all the pages script and put the `<link rel="stylesheet" href="....css">` **last** So there's no FOUC.

That could be the reason it still works. Maybe not accurate but I would say it "Hold everything and display all once ready" to make it easy. I'm not using Skeleton theming yet but normally `class="dark"` from tw will turn any bg-* color (outside SvelteKit). You can see the class in a moment before the whole page appear.

Additionally, on Astro (not sure on Vite), since it paint progressively, I can see the time difference onMount() after the page load because the svelte client js is loaded before it can run.

Anyway this is outside the scope of this PR and the #2259 issue.

---

**TLDR;** It is enough to say `onMount()` should work (for SvelteKit SSR disabled). In the case if you or maintainer want to mention outside SvelteKit, it is useful to tell user to ensure `autoModeWatcher()` to run immediately e.g. `<script>` inside head. But I can understand not the focus as of now.

**Comment by jwatts777** [comment]
> @jwatts777 you are right that it works on your link. I was just guessing from the code.
> 
> Being curious how SvelteKit made it, I notice they:
> 
> 1. Let the head empty (see the original html downloaded) and use `Promise.all` at the end of the body.
> 2. Inject `<link rel="modulepreload" as="script" crossorigin="" href="...*.js">` to load all the pages script and put the `<link rel="stylesheet" href="...*.css">` **last** So there's no FOUC.

This is more or less what I arrived at after going through a couple of github issues and checking into the Sveltekit repo. This is not even the surface of `svelte:head` issues w/ Sveltekit (do a google search for svelte head duplication, etc.).

> Anyway this is outside the scope of this PR and the #2259 issue.

Definitely is, but it doesn't hurt to map out the docs with more extensive use/support. Will be adding a note that these are *not* supported, but *should work*.

> **TLDR;** It is enough to say `onMount()` should work (for SvelteKit SSR disabled). In the case if you or maintainer want to mention outside SvelteKit, it is useful to tell user to ensure `autoModeWatcher()` to run immediately e.g. `<script>` inside head. But I can understand it is not the focus as of now.

I think I will implement this, and call it Other Frameworks. This is the same conclusion that I arrived at, and I was going to call this issue *Lightswitch workarounds*. Might be better to just add it in here, rather than making another issue.

**Comment by CallMeLaNN** [comment]
Yeah ok. I was under impression that the non SSR SvelteKit is the same as any other SSR in general so I raised the concern of FOUC but turn out SvelteKit make it easy and it's not an issue and makes my concern out of scope from original issue.

---

Btw, I'm trying to integrate Skeleton with my project. I saw Dark mode improvement discussion and I have simplified version of LightSwitch, it's theming boilerplate that should not be in a project. Thinking why no lib for this. Too bad I just found this lib. I have other related suggestions I'll discuss once I managed to integrate.

**Comment by jwatts777** [comment]
> Yeah ok. I was under impression that the non SSR SvelteKit is the same as any other SSR in general so I raised the concern of FOUC but turn out SvelteKit make it easy and it's not an issue and makes my concern out of scope from original issue.

It's ok, you managed to find a typo, which is always good. This issue was primarily because No-SSR Sveltekit is a unusual and unrecommended implementation.

> Btw, I'm trying to integrate Skeleton with my project. I saw Dark mode improvement discussion and I have simplified version of LightSwitch, it's theming boilerplate that should not be in a project. Thinking why no lib for this. Too bad I just found this lib. I have other related suggestions I'll discuss once I managed to integrate.

Try joining the Discord if you want more help, the community is more active there than here. `support-forum` would be the best place.
#2223 is a good reference for design policy, and gives the roadmap.

**Comment by endigo9740** [comment]
Just FYI I've been working through PR audits all day. We've down to a handful of the most broad or the issues that need the most testing on my end. Given this, expect to see some movement on this tomorrow afternoon immediately following the new Skeleton release.

Thanks for your patience sin the meantime @jwatts777

**Comment by endigo9740** [comment]
@jwatts777 I understand the issue you're trying to tackle here. Your additions seems like a good fix for the problem.

However, note that I've been meaning to overhaul this documentation page for ages. I've ahead and co-opted your PR to handle this. I hope you don't mind. I'd recommend you read over and ensure all is well. If you give the thumbs up then this is ready to merge!

**Comment by endigo9740** [comment]
We'll go ahead and merge this, but if you spot any issues feel free to send a follow up PR. Thanks!
